test/syscalls/linux/kill.cc

```diff
@@ -103,6 +103,138 @@ TEST(KillTest, CanKillAllPIDs) {
+      child_pid = clone(
+          KillCannotTermReaper,
+          reinterpret_cast<void*>(stack + stack_size),
+          CLONE_NEWPID | SIGCHLD,
+          /* arg = */ nullptr),
+      SyscallSucceeds());
+
+  int status;
+  ASSERT_THAT(waitpid(child_pid, &status, 0),
+              SyscallSucceedsWithValue(child_pid));
+  EXPECT_TRUE(WIFEXITED(status) && WEXITSTATUS(status) == 0)
+      << "status = " << status;
+}
+
+static int KillCanTermNonReaper(void *args) {
+  int status;
+  if (getpid() != REAPER_PID) {
+    exit(-1);
+  }
+  pid_t pid = fork();
+  if (pid == 0) {
+    while (true) {
+      pause();
+    }
+  }
+  if (kill(pid, SIGTERM) != 0) {
+    exit(-1);
+  }
+  if (RetryEINTR(waitpid)(pid, &status, 0) != pid) {
+      exit(-1);
+  }
+  // NonReaper process killed by SIGTERM.
+  if (WIFSIGNALED(status) && WTERMSIG(status) == SIGTERM) {
+      exit(0);
+  }
+  exit(-1);
+}
+
+// Nonreaper process of pid namespace can recieve blockable
+// signal with SIG_DFL action, such as SIGTERM.
+TEST(KillTest, CanTermNonReaper) {
+  int stack_size = kPageSize*4;
+  char* stack = reinterpret_cast<char*>(malloc(stack_size));
+  int child_pid;
+  // Run real case under new pid namespace.
+  ASSERT_THAT(
+      child_pid = clone(
+          KillCanTermNonReaper,
+          reinterpret_cast<void*>(stack + stack_size),
+          CLONE_NEWPID | SIGCHLD,
+          /* arg = */ nullptr),
+      SyscallSucceeds());
+
+  int status;
+  ASSERT_THAT(waitpid(child_pid, &status, 0),
+              SyscallSucceedsWithValue(child_pid));
+  EXPECT_TRUE(WIFEXITED(status) && WEXITSTATUS(status) == 0)
+      << "status = " << status;
+}
+
+static bool ReaperActed = false;
+
+static void ReaperSigHandler(int sig, siginfo_t* info, void* context) { ReaperActed = true; }
+
+static int KillCanTermReaperAction(void *args) {
+  if (getpid() != REAPER_PID) {
+    exit(-1);
+  }
+  struct sigaction sa;
+  sa.sa_sigaction = ReaperSigHandler;
+  if (sigaction(SIGTERM, &sa, nullptr) != 0) {
+    exit(-1);
+  }
+  if (kill(REAPER_PID, SIGTERM) != 0) {
+    exit(-1);
+  }
+  if (kill(REAPER_PID, 0) != 0) {
+    exit(-1);
+  }
+  // Reaper Process have process SIGTERM.
+  if (!ReaperActed) {
+    exit(-1);
+  }
+  exit(0);
+}
+
+// Reaper process of pid namespace can recieve blockable
+// signal with not SIG_DFL action, such as SIGTERM.
+TEST(KillTest, CanTermReaperAction) {
+  int stack_size = kPageSize*4;
+  char* stack = reinterpret_cast<char*>(malloc(stack_size));
+  int child_pid;
+  // Run real case under new pid namespace.
+  ASSERT_THAT(
+      child_pid = clone(
+          KillCanTermReaperAction,
+          reinterpret_cast<void*>(stack + stack_size),
+          CLONE_NEWPID | SIGCHLD,
+          /* arg = */ nullptr),
+      SyscallSucceeds());
+
+  int status;
+  ASSERT_THAT(waitpid(child_pid, &status, 0),
+              SyscallSucceedsWithValue(child_pid));
+  EXPECT_TRUE(WIFEXITED(status) && WEXITSTATUS(status) == 0)
+      << "status = " << status;
+}
+
```

Comment:
The body of these tests are all the same, except for the function that is passed to clone().

Can we use a paramaterized test to remove the duplication?

I think it will look similar to these wait() tests:
https://github.com/google/gvisor/blob/38de91b028639ef5f4a4c8874b3ee23503fd2f3a/test/syscalls/linux/wait.cc#L734

Reply:
Thanks, I have rewrite the case to paramaterized.